samza-api/src/main/java/org/apache/samza/util/BlockingEnvelopeMap.java

```diff
@@ -241,6 +243,19 @@ protected boolean isAtHead(SystemStreamPartition systemStreamPartition) {
     return getNumMessagesInQueue(systemStreamPartition) == 0 && isAtHead != null && isAtHead.equals(true);
   }
 
+  protected void setFailureCause(Throwable throwable) {
+    this.failureCause = throwable;
+  }
+
+  protected boolean isConsumerFailed() {
```

Comment:
`boolean` return type methods shouldn't throw an exception. Maybe check for exception directly and rethrow.